### DISCUSSION TITLE
How are Bundled pilot proxies renewed?

### DISCUSSION THREAD

**Discussion by sfayer** [DIRACGrid/DIRAC#7842] [Q&A]
Hi,

We're having a lot of problems with proxy expiry in longer jobs... One of the things I'm looking at is the (now default due to tokens I guess) BundleProxy mode where the pilot proxy is included in the job wrapper: Which component is responsible for renewing that proxy once the job is running (if any)?

Regards,
Simon

**Comment by atsareg** [comment]
The payload proxy is renewed by the InProcess/SingularityComputingElement components. The method used to monitor the payload proxy and to renew it as necessary is ComputingElement._monitorProxy().
If you see expired proxy problems, most likely it is because the pilots spent too much time waiting in the local queue and their rather long proxies are still not enough and they are not renewed.

**Comment by sfayer** [comment]
Hi,

Isn't that the payload proxy rather than the pilot proxy? It's the latter we're having trouble with.
(The payload proxy renewal is also broken, but as that's dependent on the pilot proxy, I'm trying to fix things in the correct sequence!)

On the old system the pilot proxy was renewed via the grid CE, but I haven't seen/found an equivalent for that in the BundledProxy mode...

Regards,
Simon

**Comment by aldbr** [comment]
IIUC, the pilot proxy is bundled with the pilot files in: 
https://github.com/DIRACGrid/DIRAC/blob/b5bc25cb2d04ac10e3998e7dd9055f38444f65cf/src/DIRAC/WorkloadManagementSystem/Utilities/PilotWrapper.py#L371


This method is called by the `SiteDirector`:
https://github.com/DIRACGrid/DIRAC/blob/b5bc25cb2d04ac10e3998e7dd9055f38444f65cf/src/DIRAC/WorkloadManagementSystem/Agent/SiteDirector.py#L692

The proxy used is the one attached to the CE:
https://github.com/DIRACGrid/DIRAC/blob/b5bc25cb2d04ac10e3998e7dd9055f38444f65cf/src/DIRAC/WorkloadManagementSystem/Agent/SiteDirector.py#L449

And this proxy attached to the CE is initialized in:
https://github.com/DIRACGrid/DIRAC/blob/b5bc25cb2d04ac10e3998e7dd9055f38444f65cf/src/DIRAC/WorkloadManagementSystem/Agent/SiteDirector.py#L891-L920

Where `proxyMinimumRequiredValidity` should be at least 1 day (unless `queueCPUTime` is negative, which would be unusual):
https://github.com/DIRACGrid/DIRAC/blob/b5bc25cb2d04ac10e3998e7dd9055f38444f65cf/src/DIRAC/WorkloadManagementSystem/Agent/SiteDirector.py#L318-L319


I might be wrong, but I don't think there is any mechanism in place to renew the pilot proxy once it is submitted to the CE.
It means that if `queueCpuTime` is 0 and (pilot stays hours in the queue and then tries to fetch jobs after it is launched) > 86400 seconds, then the proxy expires.

Is that your problem?

Because @marianne013 also mentioned the following issue:
> there's a bunch of pilot jobs [...] that basically failed immediately with "proxy expired", though at least one job sent to a different CE at the same site is still hanging in the queue.

**Reply by sfayer** [reply]
Thanks, yes the problem is that jobs can run for up to 2 days and after 1 day of running (assuming the queuing time was short) the pilot proxy expires and everything breaks.

I think there may be an additional problem with the _user_ proxy renewal, but there are so many errors caused by the pilot proxy expiring, it isn't clear if that's independent or not yet.

Are there any plans to add a pilot proxy renewal mechanism to the pilot (say in the JobAgent or something)?

Regards,
Simon
